fuel-gson/src/main/kotlin/com/github/kittinunf/fuel/gson/FuelGson.kt

```diff
@@ -21,6 +21,16 @@ import java.io.Reader
 inline fun <reified T : Any> Request.responseObject(noinline handler: ResponseResultHandler<T>) =
     response(gsonDeserializer(), handler)
 
+/**
+ * Asynchronously gets a response object of [T] wrapped in [Result] via [handler] using custom [Gson] instance
+ *
+ * @param gson [Gson] custom Gson deserializer instance
+ * @param handler [ResponseResultHandler<T>] the handler that is called upon success
+ * @return [CancellableRequest] request that can be cancelled
+ */
+inline fun <reified T : Any> Request.responseObject(gson: Gson, noinline handler: ResponseResultHandler<T>) =
```

Comment:
While we don't have anything particularly to against this, this obviously also dose the job. But what do you think about what we have done in the `fuel-moshi` extension? 

Here is the link to the master version 

https://github.com/kittinunf/fuel/blob/master/fuel-moshi/src/main/kotlin/com/github/kittinunf/fuel/moshi/FuelMoshi.kt#L15

In short, in Moshi extension, we just expose the `Moshi.Builder()` directly. What do you think about it? I think it might be just *easier* and *consistent* to do the same thing?

Reply:
Isn't `Moshi.Builder()` mutable? I'm afraid users can end messing things up, ex: you may want to decode same type from different APIs with different formats, I usually prefer to not expose unsafe APIs, but completely support you about consistency

Of course both options may be provided - shared mutable builder and explicit argument.

Also users may fallback to using `gsonDeserializer(Gson)` directly, in this case that function is not needed.

This is how my temporary solution currently looks:
```kotlin
inline fun <reified T : Any> Gson.deserializer(): ResponseDeserializable<T> =
    object : ResponseDeserializable<T> {
        override fun deserialize(reader: Reader): T? = fromJson<T>(reader, object : TypeToken<T>() {}.type)
    }

Fuel
    .get("path/to/my/service")
    .timeoutRead(timeoutMillis)
    .responseObject(gson.deserializer<List<SavedUserFeedEvent>>()) { result ->
        Either.run {
            result.fold(success = { cb(right(it)) }, failure = { cb(left(it)) })
        }
    }
```

Reply:
That's true, I think you are right. While we are exposing the mutable builder, it provides the ease of use (with the power of users can mess things up of course). But what you said is correct about exposing the safe API is usually a better choice.

I think it depends on how much compromisation we would love to make. I think maybe exposing 2 APIs options also makes sense and giving flexibility.

As you are obviously one of the users! 😄, WDYT? In whatever way, we come to the conclusion, I will open a PR for `moshi` to standardize this.

Reply:
I _vote_ for having explicit arguments for `responseObject()` functions and this will be consistent with [FuelJackson](https://github.com/kittinunf/fuel/blob/4fd0722c90d1cddfed5593fea53f502faadf7a85/fuel-jackson/src/main/kotlin/com/github/kittinunf/fuel/jackson/FuelJackson.kt) implementation.

I would strongly discourage using shared builder, but ... :) - to keep it consistent - or need to add it to `fuel-gson` also, or deprecate it from other implementations - which is a _breaking change_: so this is more a decision for project owner, my view on the project is very limited.

We can keep it as is for now and it may be added later, when somebody request it: not giving it now - is _ok_ ... but giving it now and getting it back later - will be a breaking change

Reply:
I think it's good to:

- deprecate shared builder (don't remove, yet), add default builder options to both (or all integrations)
- add function call time builder argument that defaults to a **new** builder using the **default** options.

Reply:
Let's do that then? @lostintime would you mind doing us a favor of providing both options? We will be taking care of some other extensions.

Reply:
Sure, (was a bit busy these days :)

So, just to confirm - do you want me to add a `val defaultGson = GsonBuilder()` instance and mark it as deprecated from the beginning? or just add default value to `gsonDeserializer(gson = Gson())` function, and use it everywhere needed?

Reply:
> just add default value to gsonDeserializer(gson = Gson()) function, and use it everywhere needed?

Ah I meant this!

Reply:
I've moved Gson() instance to gsonDeserializer agrument.

I've also changed `gsonDeserializerOf` to:

```kotlin
inline fun <reified T : Any> gsonDeserializerOf(clazz: Class<T>) = gsonDeserializer<T>()
```
It seemed redundant to me and there was one ore `Gson` instance used, but I'm not sure this is right, or maybe add one more gson argument here also, with default value?

Reply:
*probably it would make sense if you use this method from Java code ...

Reply:
Yes, that sounds great with a default value. But I think I am happy to see this in. 🎉